src/kernels/jupyter/launcher/liveshare/hostJupyterServer.ts

```diff
@@ -34,27 +34,35 @@ import { JupyterNotebook } from '../jupyterNotebook';
 import { noop } from '../../../../platform/common/utils/misc';
 import { Cancellation } from '../../../../platform/common/cancellation';
 import { getDisplayPath } from '../../../../platform/common/platform/fs-paths';
-import { INotebookServer, IJupyterSessionManagerFactory } from '../../types';
+import { INotebookServer } from '../../types';
 import { Uri } from 'vscode';
 /* eslint-disable @typescript-eslint/no-explicit-any */
 
 @injectable()
 export class HostJupyterServer implements INotebookServer {
-    private connection: IJupyterConnection | undefined;
-    private connectPromise: Deferred<IJupyterConnection> = createDeferred<IJupyterConnection>();
     private connectionInfoDisconnectHandler: IDisposable | undefined;
     private serverExitCode: number | undefined;
     private notebooks = new Set<Promise<INotebook>>();
-    private sessionManager: JupyterSessionManager | undefined;
     private disposed = false;
     constructor(
         @inject(IAsyncDisposableRegistry) private readonly asyncRegistry: IAsyncDisposableRegistry,
-        @inject(IJupyterSessionManagerFactory) private readonly sessionManagerFactory: IJupyterSessionManagerFactory,
         @inject(IWorkspaceService) private readonly workspaceService: IWorkspaceService,
         @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly jupyterOutputChannel: IOutputChannel,
-        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        public connection: IJupyterConnection,
+        private sessionManager: JupyterSessionManager
```

Comment:
The connection property is accessed using `getConnection` and we have some code that waits for this to be initialized.
Basically as soon as this class is contructed, the first method invoked is `connect` and its impossible for `connection` to be undefined, however due to the way the class was constructed it can be undefined in types & we have code all over the place that treats it as such.

Refactored this class to be created via a factory so that it can never be nullable, as a result the code paths are much simpler now.
No more nullables.